homework.cls

```diff
@@ -138,47 +157,44 @@
 % Mathematics packages
 \RequirePackage{amsmath, amsthm, amssymb}
 \RequirePackage{mathtools}
+\RequirePackage{physics}
 
 
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 % Problem, parts, solution, and claim environments %
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 
-% Problem environment
+\RequirePackage[dvipsnames]{xcolor}
+\def\@problemColor{white}
 \ifHW@boxes
-  \RequirePackage{framed}
+  \def\@problemColor{\HW@color}
 \fi
-\theoremstyle{definition}
-\newtheorem{@problem}{\sffamily\@problemname}
-\numberwithin{equation}{@problem}
-\numberwithin{table}{@problem}
-\numberwithin{figure}{@problem}
+
+% Problem environment
+\RequirePackage{varwidth}
+\RequirePackage{tcolorbox}
+\tcbuselibrary{theorems}
+\tcbuselibrary{skins}
+
+\newtcbtheorem{prob}{\@problemname}{enhanced, colback=\@problemColor!5, colframe=\@problemColor!50!white, coltitle=black,fonttitle=\bfseries\sffamily, colbacktitle=\@problemColor!15!white, attach boxed title to top left={yshift=-2mm,xshift=-5mm}, varwidth boxed title, borderline={0.5mm}{0mm}{\@problemColor!15!white}, borderline={0.5mm}{0mm}{\@problemColor!50!white, dashed}, boxed title style={boxrule=1pt}}{prob}
 \newenvironment{problem}[1][]
 {
-  \ifHW@boxes
-    \framed
-    \vspace{-1.5ex}
-  \fi
   \def\@arg{#1}
   \ifx\@arg\empty
-    \@problem
+    \begin{prob}{}{}
   \else
-    \@problem[\normalfont\@arg\sffamily]
+    \begin{prob}{#1}{}
   \fi
-  \pdfbookmark{\@problemname\ \arabic{@problem}}{\arabic{@problem}}
+  \pdfbookmark{\@problemname\ \arabic{tcb@cnt@prob}}{\arabic{tcb@cnt@prob}}
 }
 {
-  \end@problem
-  \ifHW@boxes
-    \vspace{-1.5ex}
-    \endframed
-  \fi
+  \end{prob}
```

Comment:
I like this idea. In the past there were more primitive box options (cf. 4cca36dc4a7ecef753db867b04f6cf525e9c2e2e again). This change would be best toggled with a class option, in my opinion. I was thinking of rewriting the `problem` implementation using [`thmtools`](https://ctan.org/pkg/thmtools?lang=en); I haven't heard of `tcolorbox`. Have you used `thmtools`, and if so, do you prefer `tcolorbox`?

Reply:
I haven't used `thmtools` before, but I just browsed the documentation and it looks nice, but maybe more basic than `tcolorbox`. The documentation for [`tcolorbox`](https://www.ctan.org/pkg/tcolorbox) is _very_ extensive, which is nice, and the overall feels like a very well thought out package. I'd personally prefer to keep this with `tcolorbox` because of the extensibility, but understand that `thmtools` is perhaps simpler.

> This change would be best toggled with a class option, in my opinion.

Do you mean the entire change, so the default look stays the same? If so I'm not sure how to handle counters because they will be different when different class options are taken. Is there a way around this?

Reply:
> I'd personally prefer to keep this with `tcolorbox`

That's fine with me

> Do you mean the entire change, so the default look stays the same?

Yes, this would be ideal in my opinion. I'm just very biased toward how it looks :) although I like your boxed version as well

> Do you mean the entire change, so the default look stays the same? If so I'm not sure how to handle counters because they will be different when different class options are taken. Is there a way around this?

I don't have an elegant solution off the top of my head. One crude way would be to have separate counters for different combinations of class options.

Reply:
I'll see if I can work something out. I'm quite busy with classes right now, so it may take a bit.